### DISCUSSION TITLE
Support for Windows Containers

### DISCUSSION THREAD

**Discussion by blackliner** [community/community#9977] [Codespaces]
Are there any plans to support Windows containers in codespaces? We develop for multiple different platforms, and Windows is amongst them. 

Reason:
Our devs usually have Ubuntu as their primary OS, and we are looking for a convenient way to give them access to a Windows platform to build and test their code.

**Comment by asciimike** [comment]
@blackliner short answer is that it's something we'd like to support, but haven't dedicated time to productionize it yet. Would you mind telling me a little more about the Windows development you're doing (e.g. do you need GUIs, would you like to use Visual Studio, etc.)?

**Reply by blackliner** [reply]
We develop our products in a monorepo, hosted on github. Some of our products are available on multiple platforms, which includes Linux and Windows. To be able to reproduce e.g. build errors when compiling with `msvc`, we need to be able to use a windows environment to replicate those issues.

We are using a Win11 + WSL2 setup to cover both platforms, but would like to go next-gen and leverage cloud capabilities like codespaces.

One additional limitation I see is that devcontainers can only provide ONE statically configured devcontainer. It seems impossible to choose different work environments, like windows, ubuntu18.04, ubuntu20.04 etc.

No need for GUI or VS at the moment.

**Reply by tgross35** [reply]
I think something super barebones would be a great start - all that's really needed is a powershell interface to same setup as the Windows runners. Debugging GH actions that run on Windows is the worst (lots of guess and check to figure out the system and environment), and this would be a nice way to make it a bit easier.

**Reply by spenserblack** [reply]
>mind telling me a little more about the Windows development you're doing

As a simple example, I was in a position where I wanted to make some OS-specific tests. Specifically, snapshot tests that included absolute filepaths. If I was able to have 2 codespaces open in different tabs -- one Linux, one Windows, it would have been awesome to work with. GitHub Actions are always a workaround, but it would have been a great development experience if I could change OS just by changing tabs, without needing to trigger and wait for unnecessary Action runs. I would have been able to work on these tests much faster with a Windows codespace.

**Reply by goyalyashpal** [reply]
> Would you mind telling me a little more about the Windows development you're doing (e.g. do you need GUIs, etc.)?

The windows tools i build are based upon linux utilities. So, developing it often means testing on widely different environments, setting which on local system is not very efficient. Being able to compile and test these things using power of cloud will help a lot.

No, I don't need any more GUIs then the Codespace with VS Code already provides. 
Having support for Github Actions (https://github.com/orgs/community/discussions/3455) will further accelerate my development productivity.

**Reply by chrpai** [reply]
@asciimike I have a huge usecase.  Developers on Mac/Linux are often creating apps for Windows.  Examples are dotnet, GoLang, Electron and so on.  They often need to create an installer for their app.

The goto MSI generator tool is Windows Installer XML (WiXToolSet ... FOSS .NET Foundation App)  While WiX v4.0 is coded in dotnet, it can't actually do anything useful on Linux/MacOS due to it's dependencies on the Windows Installer APIs that only exist on Windows.

WiX v4.0 is. dotnet tool that is restored by Nuget and is merely a command line compiler.  A windows container that had dotnet installed would allow someone on Linux/Mac to build their installer in a codespace from their native device.

This would be a HUGE win. IMO.  Today I can only do this in GitHub Actions by building some on linux and some one windows. I can't actually build the installer locally before comiting.

**Reply by Moosems** [reply]
> Would you mind telling me a little more about the Windows development you're doing (e.g. do you need GUIs, would you like to use Visual Studio, etc.)?

I have been working on a project that handles a lot of strings and since they run differently on Windows, I cannot test the code without buying a new machine to fix bugs that arise when working with unicode on Windows (some chars are handled differently like `\u200b`)

**Reply by jozefizso** [reply]
We are working on .NET library which targets .NET Framework and the Windows platform. We don't need Visual Studio capabilities, but being able to spin up a dev environment which can compile the code for .NET Framework would be beneficial.

**Comment by bubukin80** [comment]
​<?php 
 ​/* 
 ​Plugin Name: Might I Recommend... 
 ​Description: Would sir care to try this post? Or perhaps this one? 
 ​Version:     1.0 
 ​Author:      Cat Rymer 
 ​License:     GPL2 
 ​License URI: https://www.gnu.org/licenses/gpl-2.0.html 
 ​*/ 
  
 ​require​ ​'similarposts.php'​; 
  
 ​class​ clr_post_recommendations_engine { 
  
 ​        ​function​ ​__construct​() { 
 ​                ​add_filter​( ​'the_content'​, ​array​(​$​this​, ​'add_recommendations'​) ); 
 ​        } 
  
 ​        ​function​ ​add_recommendations​ ( ​$​content​ ) { 
 ​                ​if​ ( ​is_single​() ) { 
 ​                        ​$​post_categories​ = ​$​this​->​current_post_details​(); 
 ​                        ​$​recommendedpost​ = ​$​this​->​grab_similar_posts​( ​$​post_categories​ ); 
 ​                        ​$​content​ = ​$​content​ . ​"<hr />"​ . ​"<p>"​ . ​"MIGHT I RECOMMEND..."​ . ​"</p>"​ . ​$​this​->​my_test​ ( ​$​recommendedpost​ ); 
 ​                } 
 ​                ​return​ ​$​content​; 
 ​        } 
  
 ​        ​//get the categories for the current post 
 ​        ​function​ ​current_post_details​() { 
 ​                ​$​postid​ = ​get_the_ID​(); 
 ​                ​$​post_categories​ = ​wp_get_post_categories​( ​$​postid​ ); 
 ​                ​return​ ​$​post_categories​; 
 ​        } 
  
 ​        ​//grab the recommended post and save it as a transient 
 ​        ​function​ ​grab_similar_posts​ ( ​$​post_categories​ ) { 
 ​                ​global​ ​$​similar_posts_finder​; 
 ​                ​$​postid​ = ​strval​( ​get_the_ID​() ); 
 ​                ​$​transient​ = ​"recommendation"​ . ​$​postid​; 
 ​                ​$​recommendedpost​ = ​get_transient​( ​$​transient​ ); 
 ​                ​if​ ( ​$​recommendedpost​ === ​false​ ) { 
 ​                        ​$​recommendedpost​ = ​$​similar_posts_finder​->​recommended_post_builder​ ( ​$​post_categories​ ); 
 ​                        ​set_transient​( ​$​transient​, ​$​recommendedpost​, ​HOUR_IN_SECONDS​ ); 
 ​                } 
 ​                ​return​ ​$​recommendedpost​; 
 ​        } 
  
 ​        ​//build the display of the recommended post 
 ​        ​function​ ​my_test​ ( ​$​recommendedpost​ ) { 
 ​                ​$​title​ = ​$​recommendedpost​->​post_name​;  
 ​                ​$​url​ = ​get_permalink​( ​$​recommendedpost​->​ID​ ); 
 ​                ​$​snippet​ = ​substr​( ​$​recommendedpost​->​post_content​, ​0​, ​100​ ); 
 ​                ​$​recommendation​ = ​"<strong>$title - </strong><br /><em> $snippet"​ . ​"..."​ . ​" <a href='$url'>(more)</a></em>"​; 
 ​                ​return​ ​$​recommendation​; 
 ​        } 
  
 ​} 
  
 ​new​ clr_post_recommendations_engine;

**Comment by pjmlp** [comment]
Codespaces currently doesn’t cater to the needs of Windows developers, most workflows are based around using .NET alongside C++, MFC, ATL, UWP/WinUI in Visual Studio proper.

While having Windows container support would be an improvement, if it is only for headless applications, or bound to the less capable Visual Studio Code tooling, then it might also not get most adoption among this demographic anyway.

**Comment by Salman-dott** [comment]
That's a difficult question for me. I am yet a beginner and not a real expert in Windows. But when I had Win key problems, I used advice from [here](https://www.auslogics.com/en/articles/winkey-not-working-on-windows-10/) so maybe you can find something that will help you too. I really hope it will work. And maybe you could help me finding out the best materials to enhance my overall computer knowledge?

**Comment by kerbymart** [comment]
It's worth noting that GitHub is now owned by Microsoft. It's interesting to observe that Codespaces currently does not offer Windows Containers.

**Comment by craiglpeters** [comment]
There are no current plans for Windows Codespaces. C++ projects can certainly be built in Linux using `gcc`, so will work in Codespaces, unless you have Windows dependencies in your project.

If you do, [Microsoft DevBox](https://azure.microsoft.com/en-us/products/dev-box/) runs on Windows VMs and supports Visual Studio C/C++

**Reply by jozefizso** [reply]
Microsoft Dev Box service is discontinued.

> Microsoft Dev Box
> Capabilities of Microsoft Dev Box are transitioning to Windows 365. Existing customers can continue using Dev Box, but new sign-ups will no longer be accepted as of November 1, 2025. [Learn more](https://go.microsoft.com/fwlink/?linkid=2338759).

**Comment by emmadesilva** [comment]
@asciimike 

> short answer is that it's something we'd like to support, but haven't dedicated time to productionize it yet. Would you mind telling me a little more about the Windows development you're doing (e.g. do you need GUIs, would you like to use Visual Studio, etc.)?

I second this request. Would be really helpful to be able to spin up a fresh Windows environment to debug and test code in clean isolated setups. Don't need a GUI, just an SSH into a Windows VM would be great.

**Reply by emmadesilva** [reply]
cc @craiglpeters I think there is definitely valid use-cases to support Windows Codespaces, perhaps by integrating Microsoft DevBoxes?

**Comment by asciimike** [comment]
Hey folks, I haven't worked at GitHub in over two years, so @ messaging me is unlikely to help out here. Sounds like @craiglpeters has pick up this torch, and has provided guidance in https://github.com/orgs/community/discussions/9977#discussioncomment-8308865.

**Reply by emmadesilva** [reply]
My bad for not checking, I apologize!

**Comment by ctjlewis** [comment]
The only reason I need the Windows Codespace is to test the project on Windows.
